XML interface introspection

keszybz opened the pull request.

keszybz commented:
This is what I came up with to easily introspect various daemons from the command line. After the preparatory work, actual changes the the individual daemons are very simple.

boucman commented:
Do you intend to use this as a build step with update-dbus-docs.py ?

That would be great

keszybz commented:
> Do you intend to use this as a build step with update-dbus-docs.py ?

Yeah, the plan would be to substitute the introspection calls with the newly added calls.

boucman commented:
a drawback to that, which I realised after my previous comment is that it won't be possible to generate documentation when cross-compiling....

That's not too much of a change, since we already have to compile systemd twice when cross-compiling, but it's a shame to have that even more deeply engraved into the way systemd is built...

keszybz commented:
>  it won't be possible to generate documentation when cross-compiling....

This would only be done occasionally, and manually, after any dbus API changes are done. I don't think there's any reason to do it at every build, since those APIs are modified a few times per year.

boucman commented:
ok, sounds good then.

poettering commented on src/import/importd.c:
```diff
@@ -1373,6 +1374,28 @@ static bool manager_check_idle(void *userdata) {
+
+        for (i = 0; i < ELEMENTSOF(impls); i++)
+                if (streq(interface_name, impls[i]->interface))
+                        goto found;
```
i'd avoid needless jumps/labels. Maybe just do `struct bus_object_implementation *found = NULL;` at the top, then assign `found` in the loop and break for it, and then do `if (!found) return log_error(…`. Much more readable I think.

keszybz replied:
It's not needless. With the jump, the function is concise and clear. With `*found`, I need to a) do artificial initialization of the pointer and b) introduce a duplicated check for error. It's one of those cases where either `for ... else` from Python or a `goto` works much better then "structured code" or whatever it's called.

poettering commented on src/import/importd.c:
```diff
@@ -1399,6 +1422,7 @@ static int help(const char *program_path) {
                "Options:"
                "  -h --help                 Show this help\n"
                "     --version              Show package version\n"
+               " --xml-introspect=INTERFACE Write XML introspection data\n"
```
this should say "bus" or "D-Bus" somewhere. Maybe change the explanation to "Write D-Bus introspection data"?

I mean, the fact that it is *dbus* introspection is much more interesting than that it is in XML

poettering replied:
hence, also change lhs to `--bus-introspect=IFACE`?

keszybz commented on src/import/importd.c:
```diff
@@ -1125,131 +1126,114 @@ static int property_get_progress(
+        "org.freedesktop.import1.Manager",
+
+        .vtable =
+        {SD_BUS_VTABLE_START(0),
```
I wanted to minimize the indentation. If a newline is used, everything get's pushed to the right...

poettering replied:
well, but now you are pushing it one character to the right... I'd just place the bracket after the `=` and then manually remove the indentation emacs wants to add on the next line. it will then follow through with the same indentation for all further lines

keszybz commented on src/import/importd.c:
```diff
@@ -1325,16 +1336,17 @@ static int manager_add_bus_objects(Manager *m) {
+                                       transfer_object.path,
+                                       transfer_object.interface,
+                                       transfer_object.vtable,
+                                       transfer_object.find, m);
```
Is this something that we'd want to export at some point?

poettering replied:
definitely something to think about. But it would basically double our public API in size, not sure i would do that?

keszybz commented:
New version:
- goto is gone
- I kept the vtables separate, so the issue of extra indentation is sidestepped
- a build target to update the docs based on exported data is added
- a few small issues found in the previous step are fixed
- a description of LogControl1 is added

(A side note: I think the LogTarget property in LogControl1 seems problematic. It doesn't make much sense outside of systemd.)

keszybz commented:
> **new alerts:**
> 
>     * 1 for Use of 'global' at module level

Oh, I fix that later. Doesn't seem to be worth pushing just for that.

keszybz commented:
systemd.systemd on azure pipelines was failing with:
`ERROR: Command cannot have '@INPUT@', since no input files were specified`

I'll update to fix that.

poettering commented on src/shared/service-util.c:
```diff
@@ -0,0 +1,75 @@
+                return log_oom();
+
+        printf("%s [OPTIONS...]\n\n"
+               "%s\n\n"
```
we started to highlight this line recently with ANSI sequences, in systemd-nspawn for example. do here too?

poettering commented on src/shared/bus-util.c:
```diff
@@ -1549,3 +1549,47 @@ int bus_message_new_method_call(
+        log_debug("Registering bus object implementation for path=%s iface=%s", impl->path, impl->interface);
+
+        for (const BusObjectVtablePair *p = impl->vtables; p->vtable; p++) {
+                if (!p->object_find)
```
not a fan of this i must say. there *is* benefit of allowing object_find also for non-fallback kind of stuff. it's a different concept.

i..e object_find() can instantiate objects on demand, and that's useful even if you want to instantiate things only one or twice. i.e. think about an object that encapsulates something we store on disk. in object_find() we could load that, and return a pointer of what we just loaded. (machined does that). doing something like that makes sense independently of declaring that their are multiple objects of that same time. Hence, I think this should be kept independent of the specification of an obejct_find parameter, even if right now we usually combine both mechanisms.

poettering commented on src/shared/bus-util.c:
```diff
@@ -1549,3 +1549,47 @@ int bus_message_new_method_call(
+                                               impl->path);
+        }
+
+        for (unsigned i = 0; impl->children[i]; i++) {
```
size_t, we iterating through memory objects here

poettering commented on src/shared/bus-util.h:
```diff
@@ -23,11 +23,31 @@ typedef enum BusTransport {
+struct BusObjectImplementation {
+        const char *path;
+        const char *interface;
+        const BusObjectVtablePair *vtables;
```
i'd do two separate fields here, one for the plain vtables and one for the fallbcak vtables, see earlier discussion

poettering commented on src/shared/bus-log-control-api.c:
```diff
@@ -107,17 +107,9 @@ static const sd_bus_vtable log_control_vtable[] = {
+        "/org/freedesktop/LogControl1",
+        "org.freedesktop.LogControl1",
+        .vtables = VTABLE_LIST( { log_control_vtable } ),
+        .children = CHILD_LIST(),
```
a pity we have to specify this here and can't just omit this when it is empty

poettering commented on src/shared/bus-util.h:
```diff
@@ -23,11 +23,31 @@ typedef enum BusTransport {
+
+#define VTABLE_LIST(...) ((const BusObjectVtablePair[]) { __VA_ARGS__, {} })
+#define CHILD_LIST(...) { __VA_ARGS__ __VA_OPT__(,) NULL }
+#define IMPLEMENTATION_LIST(...) ((const BusObjectImplementation* []) { __VA_ARGS__, NULL })
```
hmm, where is this one used?

keszybz replied:
In the `--bus-introspect` implementations.

poettering commented on src/shared/bus-util.h:
```diff
@@ -23,11 +23,31 @@ typedef enum BusTransport {
+        const char *interface;
+        const BusObjectVtablePair *vtables;
+        sd_bus_node_enumerator_t node_enumerator;
+        bool manager;
```
not used?

keszybz replied:
It's for systemd-homed. I didn't convert it yet, but it's on my todo list (along with the api docs for it...).

poettering replied:
add the bool then back in?

poettering commented on src/shared/bus-util.c:
```diff
@@ -1593,3 +1594,103 @@ int bus_add_implementation(sd_bus *bus, const BusObjectImplementation *impl, voi
+                bool allow_fallback,
+                const BusObjectImplementation* const* bus_objects) {
+
+        for (unsigned i = 0; bus_objects[i]; i++) {
```
size_t. we iterate through memory objects here... in theory there could be as many entries in the array as we have addressable memory...

poettering commented on src/shared/bus-util.c:
```diff
@@ -1593,3 +1594,103 @@ int bus_add_implementation(sd_bus *bus, const BusObjectImplementation *impl, voi
+        for (const BusObjectVtablePair *p = impl->vtables; p->vtable; p++)
+                if (!p->object_find)
+                        return true;
+        return false;
```
hmm, not a fan, see above

poettering commented on src/shared/service-util.c:
```diff
@@ -22,9 +22,11 @@ static int help(const char *program_path, const char *service, const char *descr
                "\nSee the %s for details.\n"
                , program_path
                , description
+               , bus_introspect ? " --bus-introspect=PATH      Write XML introspection data\n" : ""
```
line break this, so that one can see quickly the the description is aligned as above?

poettering commented on src/shared/bus-util.c:
```diff
@@ -1622,6 +1622,16 @@ static const BusObjectImplementation* find_implementation(
+                FILE *out,
+                const BusObjectImplementation* const* bus_objects) {
+
+        for (unsigned i = 0; bus_objects[i]; i++) {
```
size_t

poettering commented on src/core/dbus.c:
```diff
@@ -522,8 +477,139 @@ static int bus_unit_enumerate(sd_bus *bus, const char *path, void *userdata, cha
+        .vtables = VTABLE_LIST( { bus_unit_vtable,        bus_unit_find } ),
+        .node_enumerator = bus_unit_enumerate,
+        .children = CHILD_LIST(),
+};
```
what's the rationale for having these structues here instead of the per-unit-type files? i.e. why export the vtable and the find routine instead of the structure?

keszybz replied:
I started the other way around, and it seemed more messy, and in the end I reworked it to be this way. The vtable needs to be exported anyway for the code that implements `--dump-properties`.

keszybz replied:
Oh, I now remember: those functions use helpers which are defined in dbus.c. To move them to the other files would require making the helpers non-static, which in the end results in a lot added lines.

poettering commented on man/org.freedesktop.LogControl1.xml:
```diff
@@ -0,0 +1,106 @@
+      the <citerefentry><refentrytitle>syslog</refentrytitle><manvolnum>3</manvolnum></citerefentry> call.
+      </para>
+
+      <para>Note: <command>journalctl</command> option <option>-p</option>/<option>--priority</option> may be
```
suffix `--priority` with `=` like we usually do, similar `--identifier=` below

poettering commented on src/shared/bus-util.h:
```diff
@@ -23,11 +23,31 @@ typedef enum BusTransport {
+};
+
+#define VTABLE_LIST(...) ((const BusObjectVtablePair[]) { __VA_ARGS__, {} })
+#define CHILD_LIST(...) { __VA_ARGS__ __VA_OPT__(,) NULL }
```
hmm, one more thing: is there really such a big benefit of listing them children inline of the structure, while listing the vtables via an external array? 

Having one or multiple vtables is going to be the common case, while having children is the exception, hence if anything this should be reversed, no? i.e. that the vtables are listed in-line while the children are listed as array?

that said, I think it would be a lot prettier if we'd do both arrays as external arrays, no? so that the structure becomes fixed size, and we don't have to deal with zero-initialization of that field?

keszybz replied:
That's a good point. I'll rework it like that.

keszybz commented:
Updated. The biggest change is to have two sets of tables (normal and fallback). Inline arrays are replaced by a pointer to an array. This changed the interface a bit, and it seems cleaner now.
I also added another commit to add `--help` highlighting to systemd binary.

keszybz commented:
```
Cloning into 'systemd'...
fatal: unable to access 'https://salsa.debian.org/systemd-team/systemd.git/': Received HTTP code 503 from proxy after CONNECT
Cloning into 'systemd'...
fatal: unable to access 'https://salsa.debian.org/systemd-team/systemd.git/': Received HTTP code 503 from proxy after CONNECT
blame: https://salsa.debian.org/systemd-team/systemd.git#upstream-ci
badpkg: rules extract failed with exit code 128
...
```
This seems to affect all bionic-*.

keszybz commented:
Updated to fix conflict over SetType.

poettering commented on src/shared/bus-log-control-api.h:
```diff
@@ -2,7 +2,12 @@
+extern const BusObjectImplementation log_control_object;
+static inline int bus_log_control_api_register(sd_bus *bus) {
+        return bus_add_implementation(bus, &log_control_object, NULL);
+}
```
why export the variable? why not hide it away in bus-log-control-api.c and make bus_log_control_api_register() a function like any other that is not inline?

keszybz replied:
It's used directly as an argument to service_parse_argv().

poettering commented on src/import/importd.c:
```diff
@@ -1244,97 +1319,21 @@ static const sd_bus_vtable manager_vtable[] = {
+        "/org/freedesktop/import1",
+        "org.freedesktop.import1.Manager",
+        .vtables = BUS_VTABLES( manager_vtable ),
+        .children = BUS_IMPLEMENTATIONS( &transfer_object ),
```
wouldn't place the extra spaces here either...

poettering commented on src/login/logind-dbus.c:
```diff
@@ -3754,6 +3754,15 @@ const sd_bus_vtable manager_vtable[] = {
+        .vtables = BUS_VTABLES( manager_vtable ),
+        .children = BUS_IMPLEMENTATIONS( &seat_object,
+                                         &session_object,
+                                         &user_object ),
```
here neither (and everywhere else)

poettering commented on src/shared/bus-util.c:
```diff
@@ -1598,3 +1599,106 @@ int bus_add_implementation(sd_bus *bus, const BusObjectImplementation *impl, voi
+        const BusObjectImplementation *impl, *main_impl = NULL;
+        _cleanup_free_ char *s = NULL;
+        int r;
+
```
spurious empty line?

keszybz replied:
It seems to be gone in the latest version of the patchset.

poettering commented on src/shared/bus-util.c:
```diff
@@ -1598,3 +1599,106 @@ int bus_add_implementation(sd_bus *bus, const BusObjectImplementation *impl, voi
+
+        struct introspect intro = {};
+        bool is_interface = interface_name_is_valid(pattern);
+
```
wouldn't it be nicer to have here:

```c
bool is_interface;
if (interace_name_is_valid(pattern))
        is_interface = true;
else if (object_path_is_valid(pattern))
        is_interface = false;
else 
        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Specified pattern is neither valid interface name nor object path, refusing: %s", pattern);
```

keszybz replied:
Well, that's what it comes down to anyway... After all, hopefully there are not objects in the table with an invalid name or path.

poettering commented on src/locale/localed.c:
```diff
@@ -757,7 +763,8 @@ static int run(int argc, char *argv[]) {
                                "Manage system locale settings and key mappings.",
-                               NULL,
+                               BUS_IMPLEMENTATIONS(&manager_object,
+                                                   &log_control_object),
```
see, here you have no extra space after the ( and before the )...

poettering commented on src/shared/service-util.c:
```diff
@@ -23,10 +23,13 @@ static int help(const char *program_path, const char *service, const char *descr
                , ansi_highlight(), description, ansi_normal()
                , ansi_underline(), ansi_normal()
+               , bus_introspect ?
+               "     --bus-introspect=PATH Write XML introspection data\n" : ""
```
"Write D-Bus XML introspection data"?

poettering commented on src/shared/service-util.c:
```diff
@@ -36,15 +39,19 @@ static int help(const char *program_path, const char *service, const char *descr
                 { "help",           no_argument,       NULL, 'h'                },
                 { "version",        no_argument,       NULL, ARG_VERSION        },
+                { bus_objects ? "bus-introspect" : NULL,
+                                    required_argument, NULL, ARG_BUS_INTROSPECT },
```
hmm, i'd leave the table constant, but then just output a nice error message if anyone tries it on a service that doesn't have this. mutable option tables are weird...

keszybz commented:
Updated with the suggested changes except where discussed above.
I also added a commit to port homed over (so the .manager field gets used ;)), and a stub page for org.freedesktop.home1 interface. A tiny commit on top adds SetType() to the docs.

keszybz commented:
I'm looking into the CI failure right now. It's in TEST-46-HOMED, but it might be just a fluke.

keszybz commented:
Oops, one of the vtables in homed was busted. Should be fixed now.

keszybz commented:
bionic-amd64 is failing with timeouts in fuzzers, looks unrelated.
bionic-i386 fails with a timeout during installation, looks unrelated too.
I'll merge this by hand to resolve the conflict.

poettering commented:
lgtm, but needs a rebase
